### ISSUE TITLE
[Bug]: Worker doesn't always generate files

### ISSUE THREAD

**Issue by mikelikespie** [aspect-build/rules_ts#334]
### What happened?

I am converting a project to using bazel and using ts_project. It has mostly worked so far, except there are some rules where I get issues like

```
ERROR: /foo/bar/BUILD.bazel:74:20: output 'foo/bar/baz.js' was not created
ERROR: /foo/bar/BUILD.bazel:74:20: output 'foo/bar/baz.js.map' was not created
ERROR: /foo/bar/BUILD.bazel:74:20: output 'foo/bar/baz.d.ts' was not created
```

However, if I run it with `--strategy=TsProject=darwin-sandbox` it works fine (but slow). 

it also seems to tempoarily fix it when I do `--worker_quit_after_build` on subsequent run. So I think there's a bug where worker gets into a bad state.

When I run `--worker_debug` and get into the state, I just see lines like this get created

```
arguments: --project foo/tsconfig.json --outDir foo/bar --rootDir foo/bar --declarationDir foo/bar
# Finished the work
# Beginning new work
arguments: --project foo/tsconfig.json --outDir foo/bar --rootDir foo/bar --declarationDir foo/bar
# Finished the work
# Beginning new work
arguments: --project foo/tsconfig.json --outDir foo/bar --rootDir foo/bar --declarationDir foo/bar
# Finished the work
```
(that would be for 3 runs)

so it looks like not much is happening when it gets invoked. Not sure how i can debug it further.

This also only happens to _some_ libraries, but not all

This ts_project is using a tsconfig from a parent directory,

### Version

Development (host) and target OS/architectures:

Output of `bazel --version`:

bazel 6.0.0-homebrew

Version of the Aspect rules, or other relevant rules from your
`WORKSPACE` or `MODULE.bazel` file:

```
http_archive(
    name = "aspect_rules_js",
    sha256 = "1aa0ab76d1f9520bb8993e2d84f82da2a9c87da1e6e8d121dbb4c857a292c2cd",
    strip_prefix = "rules_js-1.20.1",
    url = "https://github.com/aspect-build/rules_js/releases/download/v1.20.1/rules_js-v1.20.1.tar.gz",
)

http_archive(
    name = "aspect_rules_ts",
    sha256 = "02480b6a1cd12516edf364e678412e9da10445fe3f1070c014ac75e922c969ea",
    strip_prefix = "rules_ts-1.3.1",
    url = "https://github.com/aspect-build/rules_ts/releases/download/v1.3.1/rules_ts-v1.3.1.tar.gz",
)

http_archive(
    name = "aspect_bazel_lib",
    sha256 = "2518c757715d4f5fc7cc7e0a68742dd1155eaafc78fb9196b8a18e13a738cea2",
    strip_prefix = "bazel-lib-1.28.0",
    url = "https://github.com/aspect-build/bazel-lib/releases/download/v1.28.0/bazel-lib-v1.28.0.tar.gz",
)
```

Language(s) and/or frameworks involved:

typescript.

### How to reproduce

```shell
I was unable to reproduce this outside of monorepo at work. I tried a couple different combinations in new repo. Happy to log out more info to help debug though
```

### Fund our work

- [ ] Sponsor our open source work by donating a [bug bounty](https://opencollective.com/aspect-build/)

**Comment by gregmagolan** [comment]
Thanks for the bug report. Could you please try to gather some more diagnostics as described in https://docs.aspect.build/rules/aspect_rules_ts/docs/troubleshooting. That will help us track down the issue. Without a reproduction diagnostics & logs are the next best thing.

**Comment by mikelikespie** [comment]
Hi @gregmagolan. Thanks for the really fast response. I was actually able to isolate it and make it reproducable in a different workspace. Here's a zip of the workspace you can try to repro it with (I'm on a mac, so maybe that is related).
[build_fail.zip](https://github.com/aspect-build/rules_ts/files/10905985/build_fail.zip)

To repro it you can either do 

```
bazel build --jobs=1 --worker_verbose  //frontend/src:runtimeRetoolConstants //frontend/src:retoolColors
```

or 

```
bazel build --jobs=1 --worker_verbose //frontend/src:retoolColors  //frontend/src:runtimeRetoolConstants
```

after cleaning.

The interesting thing the first target it builds always passes, and second one always fails. (so in the latter one, `runtimeRetoolConstants` fails)

Here are teh diagnostics for 
```
bazel build --jobs=1 --worker_verbose  //frontend/src:runtimeRetoolConstants //frontend/src:retoolColors
```

console output

```
INFO: Analyzed 2 targets (46 packages loaded, 2649 targets configured).
INFO: Found 2 targets...
INFO: Created new non-sandboxed TsProject worker (id 1), logging to /private/var/tmp/_bazel_mike/fc7921c2f2f29bb777077631408b461c/bazel-workers/worker-1-TsProject.log
INFO: From Compiling TypeScript project @//frontend/src:runtimeRetoolConstants [tsc -p frontend/tsconfig.json]:
I/O Read time: 5.430872321128845
Parse time: 153.69387567043304
Program time: 176.1392079591751
Bind time: 60.773706912994385
Check time: 0.2910832166671753
printTime time: 4.438542008399963
Emit time: 4.914039850234985
transformTime time: 1.6895430088043213
Source Map time: 0.07441592216491699
commentTime time: 0.11420798301696777
I/O Write time: 0.6417078971862793
TsProject getProgram time: 0.22120797634124756
TsProject emit time: 0.24641692638397217
TsProject diagnostics time: 0.4687080383300781
INFO: From Compiling TypeScript project @//frontend/src:retoolColors [tsc -p frontend/tsconfig.json]:
TsProject applyArgs time: 0.05604195594787598
TsProject invalidate time: 0.45508408546447754
TsProject flushWatchEvents time: 0.9855419397354126
Program time: 0.47058308124542236
TsProject getProgram time: 2.126917004585266
TsProject emit time: 0.0030829906463623047
transformTime time: 0.008166074752807617
TsProject diagnostics time: 0.08087503910064697
ERROR: /Users/mike/projects/build_fail/frontend/src/BUILD.bazel:10:20: output 'frontend/src/retoolColors.js' was not created
ERROR: /Users/mike/projects/build_fail/frontend/src/BUILD.bazel:10:20: output 'frontend/src/retoolColors.js.map' was not created
ERROR: /Users/mike/projects/build_fail/frontend/src/BUILD.bazel:10:20: output 'frontend/src/retoolColors.d.ts' was not created
ERROR: /Users/mike/projects/build_fail/frontend/src/BUILD.bazel:10:20: Compiling TypeScript project @//frontend/src:retoolColors [tsc -p frontend/tsconfig.json] failed: not all outputs were created or valid
INFO: Elapsed time: 7.957s, Critical Path: 1.99s
INFO: 22 processes: 11 internal, 5 darwin-sandbox, 4 local, 2 worker.
FAILED: Build did NOT complete successfully
```

here is the worker log for that one run

[worker-1-TsProject.log.zip](https://github.com/aspect-build/rules_ts/files/10905996/worker-1-TsProject.log.zip)


I also searched in the execroot and the files are actually never emitted.

**Comment by thesayyn** [comment]
Update on this; this is a general problem in worker mode having more than one `ts_project` within the same package that has the same tsconfig behaves horribly.

**Comment by thesayyn** [comment]
#309 is suffering from the same issue.

**Comment by Jdban** [comment]
> this is a general problem in worker mode having more than one `ts_project` within the same package that has the same tsconfig behaves horribly.

This makes sense. It's happening to us in our work monorepo, and that is our setup too. 

Our symptom was that after 3 builds (always 3), it would fail until a `bazel shutdown` or a `bazel clean`. Our error made no sense 
`tsc tried to read file (path) that wasn't an input to it` when it clearly was (since it built correctly 2x prior, and the build file was not complicated).

**Comment by jackvincentnz** [comment]
> Update on this; this is a general problem in worker mode having more than one `ts_project` within the same package that has the same tsconfig behaves horribly.

I am seeing the `tsc tried to read file (path) that wasn't an input to it` error for a package that has a single `ts_project` target in it. So I don't think that is the main issue. When I set `supports_workers = False,` I no longer see the issue. 

Let me know if you would like me to create a minimum reproduction of this or if it should be a different issue.

I am mostly copying the https://github.com/aspect-build/bazel-examples/tree/main/react-cra example, but have only got the srcs `ts_project` and have moved the cra into a subdirectory of the repo.

e.g. 
```
# //apps/react-cra/src/BUILD.bazel
ts_project(
    name = "src_ts",
    srcs = glob(
        include = SRC_PATTERNS,
        exclude = TEST_PATTERNS,
    ),
    declaration = True,
    resolve_json_module = True,
    #    supports_workers = False,
    tsconfig = "//apps/create-react-app:tsconfig",
    deps = [
        "//:node_modules/react",
        "//:node_modules/react-dom",
        "//:node_modules/web-vitals",
    ],
)
```

```
Compiling TypeScript project @//apps/create-react-app/src:src_ts [tsc -p apps/create-react-app/tsconfig.json] 
failed: (Exit 1): tsc_worker.sh failed: error executing command (from target //apps/create-react-app/src:src_ts) 
bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/npm_typescript/tsc_worker.sh 
@bazel-out/darwin-fastbuild/bin/apps/create-react-app/src/App.js-0.params

Error: tsc tried to read file (/bazel-out/darwin-fastbuild/bin/apps/create-react-app/src/app.tsx) 
that wasn't an input to it. at Object.readFile 
(/.../bin/external/npm_typescript/tsc_worker.sh.runfiles/npm_typescript/ts_project_worker.js:763:19)

```

I see the above error when running `bazel build //apps/create-react-app/src:src_ts` several times with minor changes to the `app.tsx`

**Comment by gregjacobs** [comment]
Just to add a (very) little bit of info to this thread: I'm still using rules_ts@1.0.2 because of this issue. So the issue must have been introduced somewhere between 1.0.2 and 1.3.1 :) I can try to do some more testing to see which version in particular introduced the problem if that's helpful.

**Comment by agirorn** [comment]
I have tried to have separate tsconfig files and ts_config targets in the same package, which does not solve the issue for me.

I have a test branch [her](https://github.com/agirorn/monorepo/blob/bug/aspect-build-issue-334/packages/npm/one/tests/integration/dao/OneDAO.test.ts) where this happens every time after the first failed to compile.
This started happening after I added the packages/npm/one/tests/integration/dao/OneDAO.test.ts and was working fine when I only had the /packages/npm/one/tests/integration/one.test.ts file in the test.

This jest test works fine all the time like this
```ts
test("test", () => {
  // expect(!!something).toBe(true); // BAD CODE
  expect("text").toBe("text");
});
```

Then I add in the buggy code that causes this error in tsc "typescript: Cannot find name 'something'."
```ts
test("test", () => {
  expect(!!something).toBe(true); // BAD CODE
  expect("text").toBe("text");
});
```
After that, I get the "was not created" errors, and even after I return the code back to a working state, I'm still unable to build it using bazel except if I either run `bazel clean` first or add `--strategy=TsProject=darwin-sandbox` to this build command.
```ts
test("test", () => {
  // expect(!!something).toBe(true); // comment in this line
  expect("text").toBe("text");
});
```

I would dig deeper, but I don't know where to start to debug and investigate the Bazel internals or the rules.
